Language/Java/Transformer.hs

```diff
@@ -0,0 +1,402 @@
+  analyze scope (MethodCall _ idnt args) = MethodCall Nothing idnt (map (analyze scope) args)
+  analyze scope (PrimaryMethodCall expr refTypes idnt args) = PrimaryMethodCall (analyze scope expr) refTypes idnt (map (analyze scope) args)
+  analyze scope (SuperMethodCall refTypes idnt args) = SuperMethodCall refTypes idnt (map (analyze scope) args)
+  analyze scope (ClassMethodCall name refTypes idnt args) = ClassMethodCall name refTypes idnt (map (analyze scope) args)
+  analyze scope (TypeMethodCall name refTypes idnt args) = TypeMethodCall name refTypes idnt (map (analyze scope) args)
+
+classifyName :: Ident -> IdentCollection -> Name -> ClassifiedName
+classifyName idnt ic =
+  if isIdent idnt ic then ExpressionName else TypeName
+
+-- helper functions
+
+identFromVarDeclId :: VarDeclId -> Ident
+identFromVarDeclId (VarId idnt) = idnt
+identFromVarDeclId (VarDeclArray _ varDeclId) = identFromVarDeclId varDeclId
+
+-- class fields from ClassDecl
+
+fieldsFromClassBody :: ClassBody p -> [Ident]
+fieldsFromClassBody (ClassBody decls) = identsfromDecls decls
+
+fieldsFromEnumBody :: EnumBody p -> [Ident]
+fieldsFromEnumBody (EnumBody _ decls) = identsfromDecls decls
+
+identsfromDecls :: [Decl p] -> [Ident]
+identsfromDecls decls = do
+  map
+    (identFromVarDeclId . (\(VarDecl _ varId _) -> varId))
+    ( concatMap
+        ( \case
+            (MemberDecl (FieldDecl _ _ _ vars)) -> vars
+            _ -> []
+        )
+        decls
+    )
+
+-- | idnt of exception parameter from Catch
+exceptionParameter :: Catch p -> [Ident]
+exceptionParameter (Catch (FormalParam _ _ _ _ vardeclId) _) = [identFromVarDeclId vardeclId]
+
+-- | idents of formal parameters of method and constructor declaration
+methodParameters :: MemberDecl p -> [Ident]
+methodParameters (MethodDecl _ _ _ _ _ params _ _ _) = identsFromParams params
+methodParameters (ConstructorDecl _ _ _ _ params _ _) = identsFromParams params
+methodParameters _ = []
+
+identsFromParams :: [FormalParam p] -> [Ident]
+identsFromParams = map (identFromVarDeclId . (\(FormalParam _ _ _ _ vardeclId) -> vardeclId))
+
+-- | idents of local variables from basicFor, enhancedFor and tryWithResource
+localVarsInTryResources :: [TryResource p] -> [Ident]
+localVarsInTryResources = mapMaybe identFromTryResource
+  where
+    identFromTryResource (TryResourceVarDecl (ResourceDecl _ _ vardDeclId _)) = Just (identFromVarDeclId vardDeclId)
+    identFromTryResource _ = Nothing
+
+localVarsInBasicForInit :: Maybe (ForInit p) -> [Ident]
+localVarsInBasicForInit (Just (ForLocalVars _ _ varDecls)) = map (identFromVarDeclId . (\(VarDecl _ varid _) -> varid)) varDecls
+localVarsInBasicForInit _ = []
+
+analyzeBlockStmts :: IdentCollection -> [BlockStmt Parsed] -> [BlockStmt Analyzed]
+analyzeBlockStmts __ [] = []
+analyzeBlockStmts scope (x@(LocalVars _ _ _ varDecls) : xs) =
```

Comment:
Wie ist das Scoping in Java, wenn hier mehrere Variablen auf einmal deklariert werden. Kann dann eine Variable, die weiter hinten deklariert wird, auf die Deklaration weiter vorne zugreifen. Also wenn ich zum Beispiel die folgende Zeile habe.

```java
int x = 3, y = x + 1;
```

Ist eine solche Definition valide und referenziert das `x` hier die lokale Variable zuvor? Falls das geht, wäre die aktuelle Implementierung nicht korrekt.

Reply:
👍